### PR TITLE
test: Added new test cases on Batch and Serial and batch bundle docty…

### PR DESCRIPTION
**Pull request by Prateek8848**
**Title**:
Test Cases for Stock Module Doctypes with Database Rollback and Data Integrity Validation

**Description**:
Previously, critical doctypes in the Stock module such as Serial and Batch Bundle and Batch lacked comprehensive test coverage, making them susceptible to regression issues and potential data inconsistency.
This commit introduces detailed test cases for these doctypes, ensuring accurate creation, validation, linking, and cleanup of stock-related records. The tests use realistic data and verify core workflows like bundle creation, batch status updates, and integrity checks post-deletion.

The tearDown method has been implemented to ensure database changes are rolled back after each test, maintaining test isolation and ensuring no data leakage occurs between test cases.

**Doctypes Covered**:

Serial and Batch Bundle

Batch

**Test Cases Implemented**:

test_autoname_for_naming_series_TC_SCK_271
Tests that the system raises ValidationError when naming series is required but missing.

test_autoname_for_naming_series_logic_TC_SCK_272:
Behavior when naming series is a plain string vs with #####.
Ensures autoname() generates proper names based on series.
test_autoname_for_duplicate_entry_TC_SCK_273:Simulates DuplicateEntryError by monkey-patching frappe.generate_hash.
test_make_serial_no_TC_SCK_277:Ensures serial number creation works via make_serial_no() function.
test_make_batch_no_TC_SCK_278:Validates that a new batch is successfully created for an item with batch enabled.
test_is_serial_batch_no_exists_for_serial_no_TC_SCK_279:Ensures validation error is raised when checking for a non-existent serial number in 'Maintenance' transaction.
test_is_serial_batch_no_exists_for_batch_no_TC_SCK_280:Ensures validation error is raised when checking for a non-existent batch number in 'Maintenance' transaction.
test_is_serial_batch_no_exists_for_serial_no_inward_TC_SCK_281:Verifies that no error is raised when a serial number is checked during an 'Inward' transaction.
test_is_serial_batch_no_exists_for_batch_no_inward_TC_SCK_282:Verifies that no error is raised when a batch number is checked during an 'Inward' transaction.
test_get_picked_serial_nos_TC_SCK_283:Validates that picked serial numbers for an item are returned as a list of strings.
test_get_auto_data_has_serial_no_TC_SCK_284:Ensures get_auto_data() works when has_serial_no = 1.
test_get_auto_data_has_batch_no_TC_SCK_285:That assumes a serial format even for batch-only items, which may not apply. Replace with
test_get_serial_and_batch_ledger_TC_SCK_288:
Tests combined batch and serial ledger data.
Creates necessary items, batch, and serial entries.
test_item_query_filters_TC_SCK_289:Verifies item_query results.
test_calculate_outgoing_rate_TC_SCK_290: This seems to be incomplete in your message (serial...).

**Key Enhancements**:

Introduced tearDown(self) method using frappe.db.rollback() to maintain test isolation.

All tests ensure no residual data pollutes subsequent test executions.

Improved consistency by using test utilities to simulate real-world scenarios with valid stock and batch data.

**Scenarios Covered**:

Serial and Batch Bundle creation, linkage, and deletion

Batch status transitions and edge-case handling

Proper field mapping and validation logic

Cleanup behavior and rollback integrity

Inter-doctype linkage validation (e.g., Stock Entry, Item)

**Technology Used**:

Python unittest framework

Frappe test utilities and mock data tools

Database rollback using frappe.db.rollback() in tearDown

**Linked Feature/Bug**:
N/A

**Issue ID**:
N/A

### CODE REVIEW

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 1]
```diff
@@ -7,16 +7,967 @@
...
+
+	# codecov
+	def test_autoname_for_naming_series_TC_SCK_271(self):
+		company = "_Test Indian Registered Company"
+		warehouse = "Stores - _TIRC"
+
+		from erpnext.accounts.doctype.payment_entry.test_payment_entry import create_customer, make_test_item
+
+		if not frappe.db.exists("Warehouse", warehouse):
+			warehouse = create_warehouse(warehouse, company=company)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:34]: always check if the calling method is handling the duplicate entry or not if not then only check the record before calling else no need to call database api.

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 2]
```diff
@@ -7,16 +7,967 @@
...
+				"voucher_no": dn.name,
+				"posting_date": frappe.utils.now(),
+			}
+		).insert(ignore_permissions=True)
+		serial_batch_bundle.submit()
+
+		#  Force Stock Settings to require naming series
+		frappe.db.set_value(
+			"Stock Settings", None, "set_serial_and_batch_bundle_naming_based_on_naming_series", 1
+		)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:184]: don't update the settings directly, use decorator

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 3]
```diff
@@ -7,16 +7,967 @@
...
+
+		# Reload the doc
+		serial_batch_bundle = frappe.get_doc("Serial and Batch Bundle", serial_batch_bundle.name)
+
+		#  Clear naming_series
+		serial_batch_bundle.naming_series = None
+
+		#  Assert that autoname() raises ValidationError
+		with self.assertRaises(frappe.ValidationError):
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:193]: what validation we testing here, use the msg parameter with assertRaises

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 4]
```diff
@@ -7,16 +7,967 @@
...
+				"has_batch_no": 1,
+				"naming_series": "TESTSERIES",  # no '#'
+			}
+		)
+
+		try:
+			# Call autoname
+			serial_batch_bundle.autoname()
+		except NameError as e:
+			frappe.log_error(f"autoname NameError: {e}")
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:221]: why are we logging error in the database here?

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 5]
```diff
@@ -7,16 +7,967 @@
...
+				"has_batch_no": 1,
+				"naming_series": "TESTSERIES.#####",
+			}
+		)
+
+		try:
+			# Call autoname
+			serial_batch_bundle_with_hash.autoname()
+		except NameError as e:
+			frappe.log_error(f"autoname NameError: {e}")
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:246]: wrong use of frappe.log_error

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 6]
```diff
@@ -7,16 +7,967 @@
...
+
+		# Assert that name was generated
+		self.assertTrue(serial_batch_bundle_with_hash.name.startswith("TESTSERIES"))
+
+	# codecov
+	def test_autoname_for_duplicate_entry_TC_SCK_273(self):
+		# Turn OFF the Stock Settings flag
+		frappe.db.set_value(
+			"Stock Settings", None, "set_serial_and_batch_bundle_naming_based_on_naming_series", 0
+		)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:260]: use decorator

File: erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py [hunk 7]
```diff
@@ -7,16 +7,967 @@
...
+			"warehouse": warehouse,
+			"has_serial_no": item.has_serial_no,
+			"has_batch_no": item.has_batch_no,
+			"serial_nos": serial_no,
+			"qty": 5,
+			"based_on": "FIFO",
+			"posting_date": frappe.utils.now(),
+			"posting_time": frappe.utils.now(),
+		}
+		get_serial_and_batch_ledger = get_serial_and_batch_ledger(**args)
```
Comment (aasif-patel) [inline erpnext/stock/doctype/serial_and_batch_bundle/test_serial_and_batch_bundle.py:715]: what are we testing here?

**Review by aasif-patel — CHANGES_REQUESTED** [review]
@Prateek8848 do add suggested changes in the commented line also in other test cases.
